drivers/lora/sx1276.c

```diff
@@ -0,0 +1,483 @@
+	}
+
+	return spi_write(dev_data.spi, &dev_data.spi_cfg, &tx);
+}
+
+int sx1276_read(u8_t reg_addr, u8_t *data, u8_t len)
+{
+	return sx1276_transceive(reg_addr, false, data, len);
+}
+
+int sx1276_write(u8_t reg_addr, u8_t byte)
+{
+	return sx1276_transceive(reg_addr | BIT(7), true, &byte, 1);
+}
+
+void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_write(addr, buffer[i]);
+		if (ret < 0) {
+			LOG_ERR("Unable to read address: %x", addr);
+			return;
+		}
+	}
+}
+
+void SX1276ReadBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+	u8_t regval;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_read(addr, &regval, 1);
+		if (ret < 0) {
+			LOG_ERR("Unable to read address: %x", addr);
+			return;
+		}
+
+		buffer[i] = regval;
+	}
+}
+
+void SX1276SetRfTxPower(int8_t power)
+{
+	int ret;
+	u8_t pa_config = 0;
+	u8_t pa_dac = 0;
+
+	ret = sx1276_read(SX1276_REG_PA_CONFIG, &pa_config, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read PA config");
+		return;
+	}
+
+	ret = sx1276_read(SX1276_REG_PA_DAC, &pa_dac, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read PA dac");
+		return;
+	}
+
+	pa_config = (pa_config & RF_PACONFIG_MAX_POWER_MASK) | 0x70;
+	pa_config &= RF_PACONFIG_PASELECT_MASK;
+
+#if defined CONFIG_PA_BOOST_PIN
+	pa_config |= RF_PACONFIG_PASELECT_PABOOST;
+
+	if (power > 17) {
+		pa_dac = (pa_dac & RF_PADAC_20DBM_MASK) | RF_PADAC_20DBM_ON;
+	} else {
+		pa_dac = (pa_dac & RF_PADAC_20DBM_MASK) | RF_PADAC_20DBM_OFF;
+	}
+
+	if ((pa_dac & RF_PADAC_20DBM_ON) == RF_PADAC_20DBM_ON) {
+		if (power < 5) {
+			power = 5;
+		} else if (power > 20) {
+			power = 20;
+		}
+
+		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power - 5) & 0x0F);
+	} else {
+		if (power < 2) {
+			power = 2;
+		} else if (power > 17) {
+			power = 17;
+		}
+
+		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power - 2) & 0x0F);
+	}
+#elif CONFIG_PA_RFO_PIN
+	if (power < -1) {
+		power = -1;
+	} else if (power > 14) {
+		power = 14;
+	}
+
+	pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power + 1) & 0x0F);
+#endif
+	ret = sx1276_write(SX1276_REG_PA_CONFIG, pa_config);
+	if (ret < 0) {
+		LOG_ERR("Unable to write PA config");
+		return;
+	}
+
+	ret = sx1276_write(SX1276_REG_PA_DAC, pa_dac);
+	if (ret < 0) {
+		LOG_ERR("Unable to write PA dac");
+		return;
+	}
+}
+
+static int sx1276_lora_send(struct device *dev, u8_t *data, u32_t data_len)
+{
+	Radio.SetMaxPayloadLength(MODEM_LORA, data_len);
+
+	Radio.Send(data, data_len);
+
+	return 0;
+}
+
+static void sx1276_rx_done(u8_t *payload, u16_t size, int16_t rssi, int8_t snr)
+{
+	dev_data.rx_buf = payload;
+	dev_data.rx_len = size;
+
+	k_sem_give(&dev_data.data_sem);
+}
+
+static int sx1276_lora_recv(struct device *dev, u8_t *data)
+{
+	Radio.SetMaxPayloadLength(MODEM_LORA, 255);
+	Radio.Rx(0);
+
+	/* Wait until the data arrives */
+	k_sem_take(&dev_data.data_sem, K_FOREVER);
```

Comment:
I think `sx1276_lora_recv()` needs timeout.

Reply:
@KwonTae-young Makes sense! I think the API should allow this timeout value to be specified by the application. If timeout passed is `0`, then it should block until the data arrives. What do you think?

Like below,

```
static int sx1276_lora_recv(struct device *dev, u8_t *data, u32_t timeout)
{
        int ret;

        Radio.SetMaxPayloadLength(MODEM_LORA, 255);
        Radio.Rx(0);

        if (!timeout) {
                /* Wait until the data arrives */
                k_sem_take(&dev_data.data_sem, K_FOREVER);
        } else {
                ret = k_sem_take(&dev_data.data_sem, K_MSEC(timeout));
                if (ret < 0) {
                        LOG_ERR("Receive timeout!");
                        return ret;
                }
        }

        memcpy(data, dev_data.rx_buf, dev_data.rx_len);

        return dev_data.rx_len;
}
```

Reply:
>I think the API should allow this timeout value to be specified by the application. If timeout passed is 0, then it should block until the data arrives. What do you think?

I would suggest using the already defined macros K_FOREVER and K_NO_WAIT for that purpose.

Reply:
> I would suggest using the already defined macros K_FOREVER and K_NO_WAIT for that purpose.

@alexanderwachter Okay. I think you're talking about passing `K_NO_WAIT` from the application instead of passing `0`.

Reply:
timeout works fine.
However, if the timeout occurs once in `samples/drivers/lora/receive`, it is returned and the program terminates.
```
***** Booting Zephyr OS version 2.0.99 *****              
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found  
[00:00:01.228,000] <err> sx1276: Receive timeout!
[00:00:01.228,000] <err> lora_receive: LoRa receive failed
```

I am testing with the following modifications. (timeout = 1000ms)
``` diff
diff --git a/samples/drivers/lora/receive/src/main.c b/samples/drivers/lora/receive/src/main.c
index 46824dd69e..99b6985165 100644
--- a/samples/drivers/lora/receive/src/main.c
+++ b/samples/drivers/lora/receive/src/main.c
@@ -48,11 +48,10 @@ void main(void)
                len = lora_recv(lora_dev, data, 0);
                if (len < 0) {
                        LOG_ERR("LoRa receive failed");
-                       return;
+               } else {
+                       LOG_INF("Received data: %s", log_strdup(data));
                }
 
-               LOG_INF("Received data: %s", log_strdup(data));
-
                /* Receive data at 1s interval */
                k_sleep(1000);
        }

```
```
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.228,000] <err> sx1276: Receive timeout!
[00:00:01.228,000] <err> lora_receive: LoRa receive failed
[00:00:03.229,000] <err> sx1276: Receive timeout!
[00:00:03.229,000] <err> lora_receive: LoRa receive failed
[00:00:05.230,000] <err> sx1276: Receive timeout!
[00:00:05.230,000] <err> lora_receive: LoRa receive failed
...
...
```

Reply:
@KwonTae-young This is the expected behavior. And the termination depends on the user application. The sample application I have provided just terminates once the timeout is received.

Reply:
@Mani-Sadhasivam Okay. timeout works fine.